src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2670,5 +2671,260 @@ private interface INegator<T> where T : struct
             public static Vector128<T> NegateIfNeeded(Vector128<T> equals) => ~equals;
             public static Vector256<T> NegateIfNeeded(Vector256<T> equals) => ~equals;
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void Reverse<T>(ref T elements, nuint length)
+        {
+            Debug.Assert(length > 0);
+            if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                if (Unsafe.SizeOf<T>() == sizeof(byte))
+                {
+                    ReverseSIMD(ref Unsafe.As<T, byte>(ref elements), length);
+                    return;
+                }
+                else if (Unsafe.SizeOf<T>() == sizeof(char))
+                {
+                    ReverseSIMD(ref Unsafe.As<T, short>(ref elements), length);
+                    return;
+                }
+                else if (Unsafe.SizeOf<T>() == sizeof(int))
+                {
+                    ReverseSIMD(ref Unsafe.As<T, int>(ref elements), length);
+                    return;
+                }
+                else if (Unsafe.SizeOf<T>() == sizeof(long))
+                {
+                    ReverseSIMD(ref Unsafe.As<T, long>(ref elements), length);
+                    return;
+                }
+            }
+            ReverseScalar(ref elements, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void ReverseScalar<T>(ref T elements, nuint length)
+        {
+            if (length <= 1)
+                return;
+            ref T first = ref elements;
+            ref T last = ref Unsafe.Subtract(ref Unsafe.Add(ref first, (int)length), 1);
+            do
+            {
+                T temp = first;
+                first = last;
+                last = temp;
+                first = ref Unsafe.Add(ref first, 1);
+                last = ref Unsafe.Subtract(ref last, 1);
+            } while (Unsafe.IsAddressLessThan(ref first, ref last));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void ReverseSIMD<T>(ref T elements, nuint length) where T : struct
+        {
+            if (Avx2.IsSupported && (nuint)Vector256<T>.Count * 2 <= length)
+            {
+                nuint numElements = (nuint)Vector256<T>.Count;
+                nuint numIters = (length / numElements) / 2;
```

Comment:
I imagine this will end up being at best a `(length >> cns) >> cns` rather than a direct `(length >> cns)` that you'd get if you did pre-multiply by 2 instead